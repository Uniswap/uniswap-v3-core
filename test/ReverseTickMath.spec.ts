import {ethers} from 'hardhat'
import {ReverseTickMathTest} from '../typechain/ReverseTickMathTest'

import {expect} from './shared/expect'
import snapshotGasCost from './shared/snapshotGasCost'
import {encodePrice} from './shared/utilities'

describe('ReverseTickMath', () => {
  let reverseTickMath: ReverseTickMathTest
  before('deploy test contract', async () => {
    const reverseTickMathTest = await ethers.getContractFactory('ReverseTickMathTest')
    reverseTickMath = (await reverseTickMathTest.deploy()) as ReverseTickMathTest
  })

  describe('#getTickFromPrice', () => {
    it('lowerBound = upperBound - 1', async () => {
      expect(await reverseTickMath.getTickFromPrice({_x: encodePrice(1, 1)}, 0, 1)).to.eq(0)
    })
    it('lowerBound = upperBound - 4', async () => {
      expect(await reverseTickMath.getTickFromPrice({_x: encodePrice(1, 1)}, -3, 1)).to.eq(0)
    })
    it('upperBound = lowerBound + 4', async () => {
      expect(await reverseTickMath.getTickFromPrice({_x: encodePrice(1, 1)}, 0, 4)).to.eq(0)
    })
    it('lowerBound and upper bound are both off', async () => {
      expect(await reverseTickMath.getTickFromPrice({_x: encodePrice(1, 1)}, -3, 4)).to.eq(0)
    })
    it('lowerBound and upper bound off by 128', async () => {
      expect(await reverseTickMath.getTickFromPrice({_x: encodePrice(1, 1)}, -128, 128)).to.eq(0)
    })
    it('price is below lower bound', async () => {
      expect(await reverseTickMath.getTickFromPrice({_x: encodePrice(1, 1)}, 5, 250)).to.eq(5)
    })
    it('price is above upper bound', async () => {
      expect(await reverseTickMath.getTickFromPrice({_x: encodePrice(1, 1)}, -100, -25)).to.eq(-25)
    })
    it('throws if lowerBound == upperBound', async () => {
      await expect(reverseTickMath.getTickFromPrice({_x: encodePrice(1, 1)}, 0, 0)).to.be.revertedWith(
        'ReverseTickMath::getTickFromPrice: lower bound must be less than upper bound'
      )
    })
    it('gas cost 0 iterations', async () => {
      await snapshotGasCost(reverseTickMath.getGasCostOfGetTickFromPrice({_x: encodePrice(1, 1)}, 0, 1))
    })
    it('gas cost diff of 8', async () => {
      await snapshotGasCost(reverseTickMath.getGasCostOfGetTickFromPrice({_x: encodePrice(1, 1)}, -4, 4))
    })
    it('gas cost diff of 256', async () => {
      await snapshotGasCost(reverseTickMath.getGasCostOfGetTickFromPrice({_x: encodePrice(1, 1)}, -128, 128))
    })
  })
})
