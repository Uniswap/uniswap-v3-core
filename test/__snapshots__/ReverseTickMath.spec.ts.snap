// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReverseTickMath #getTickFromPrice gas cost 0 iterations 1`] = `168`;

exports[`ReverseTickMath #getTickFromPrice gas cost diff of 8 1`] = `3317`;

exports[`ReverseTickMath #getTickFromPrice gas cost diff of 256 1`] = `8713`;

exports[`ReverseTickMath #getTickFromPrice gas cost price exactly at 0 1`] = `1045`;
